From: Chad Monroe <chad@monroe.io>
Date: Tue, 03 Mar 2026 15:45:44 -0800
Subject: [PATCH] net: ethernet: airoha: fix FOE hash bounds check for dual PPE

Hash boundary checks used the single-PPE constant PPE_SRAM_NUM_ENTRIES
(8192), rejecting valid FOE entries from the second PPE engine.
Use airoha_ppe_get_total_sram_num_entries() which scales by the actual
number of PPE instances.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/ethernet/airoha/airoha_eth.c  |    5 -
 drivers/net/ethernet/airoha/airoha_eth.h  |    1 
 drivers/net/ethernet/airoha/airoha_ppe.c  |    2 
 drivers/net/ethernet/airoha/arht_dp_api.c |   63 ++++++++------------
 4 files changed, 31 insertions(+), 40 deletions(-)

--- a/drivers/net/ethernet/airoha/airoha_eth.c
+++ b/drivers/net/ethernet/airoha/airoha_eth.c
@@ -675,9 +675,8 @@ static void airoha_get_entry_bind(u32 ms
 	u32 reason;
 	u32 hash = FIELD_GET(AIROHA_RXD4_FOE_ENTRY, msg1);
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "Hash: %u\n", hash);
-	if (hash < PPE_SRAM_NUM_ENTRIES)
-		skb_set_hash(q->skb, hash,
-				 PKT_HASH_TYPE_L4);
+	if (hash < airoha_ppe_get_total_sram_num_entries(eth->ppe))
+		skb_set_hash(q->skb, hash, PKT_HASH_TYPE_L4);
 
 	reason = FIELD_GET(AIROHA_RXD4_PPE_CPU_REASON, msg1);
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "CPU_REASON: %u\n", reason);
--- a/drivers/net/ethernet/airoha/airoha_eth.h
+++ b/drivers/net/ethernet/airoha/airoha_eth.h
@@ -815,6 +815,7 @@ int airoha_ppe_init(struct airoha_eth *e
 void airoha_ppe_deinit(struct airoha_eth *eth);
 void airoha_ppe_init_upd_mem(struct airoha_gdm_port *port);
 struct airoha_foe_entry *airoha_ppe_foe_get_entry_locked(struct airoha_ppe *ppe, u32 hash);
+u32 airoha_ppe_get_total_sram_num_entries(struct airoha_ppe *ppe);
 u32 airoha_ppe_get_total_num_entries(struct airoha_ppe *ppe);
 struct airoha_foe_entry *airoha_ppe_foe_get_entry(struct airoha_ppe *ppe,
 						  u32 hash);
--- a/drivers/net/ethernet/airoha/arht_dp_api.c
+++ b/drivers/net/ethernet/airoha/arht_dp_api.c
@@ -9384,8 +9384,8 @@ void airoha_ppe_general_bind(struct airo
 
 	hwe->ib1 = ib1;
 
-	if (hash < PPE_SRAM_NUM_ENTRIES) {
-		airoha_ppe_foe_commit_entry_ptr(ppe, hwe, hash,0);
+	if (hash < airoha_ppe_get_total_sram_num_entries(ppe)) {
+		airoha_ppe_foe_commit_entry_ptr(ppe, hwe, hash, 0);
 	}
 	
 	return;
@@ -9494,10 +9494,10 @@ void airoha_ppe_foe_flow_update_eth_offl
 	int dscp=0;
 	
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "hash at qdma_lan_tx: %u\n", hash);
-	if (hash >= PPE_SRAM_NUM_ENTRIES || !skb->l4_hash || skb->sw_hash != false) {
+	if (hash >= airoha_ppe_get_total_sram_num_entries(ppe) || !skb->l4_hash || skb->sw_hash != false) {
 		AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "Incorrect hash:%u\n", hash);
-        return;
-    }
+		return;
+	}
 
 	if(arht_multicast_handler_for_sfu(skb))
 	{
@@ -9582,11 +9582,11 @@ void airoha_ppe_foe_flow_update_pon_offl
 	int dscp=0;
 	
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "hash at qdma_wan_tx: %u\n", hash);
-	if (hash >= PPE_SRAM_NUM_ENTRIES || !skb->l4_hash || skb->sw_hash != false) {
+	if (hash >= airoha_ppe_get_total_sram_num_entries(ppe) || !skb->l4_hash || skb->sw_hash != false) {
 		AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "Incorrect hash:%u\n", hash);
-        return;
-    }
-	
+		return;
+	}
+
 	spin_lock_bh(&ppe_lock);
 	hwe = airoha_ppe_foe_get_entry_locked(ppe, hash);
 	if (!hwe)
@@ -10183,9 +10183,8 @@ EXPORT_SYMBOL(qdma_wan_tx);
 void npu_get_entry_bind(struct sk_buff *skb, u32 hash, u32 reason)
 {
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "Hash: %u\n", hash);
-	if (hash < PPE_SRAM_NUM_ENTRIES)
-		skb_set_hash(skb, hash,
-				 PKT_HASH_TYPE_L4);
+	if (hash < airoha_ppe_get_total_sram_num_entries(glb_eth->ppe))
+		skb_set_hash(skb, hash, PKT_HASH_TYPE_L4);
 
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "CPU_REASON: %u\n", reason);
 	if (reason == PPE_CPU_REASON_HIT_UNBIND_RATE_REACHED){
@@ -10214,11 +10213,11 @@ void airoha_ppe_foe_flow_update_wifi_npu
 	unsigned int dscp_tmp = 0;
 	u32 index, data;
 	AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "hash at wifi_tx: %u\n", hash);
-	if (hash >= PPE_SRAM_NUM_ENTRIES || !skb->l4_hash || skb->sw_hash != false) {
+	if (hash >= airoha_ppe_get_total_sram_num_entries(ppe) || !skb->l4_hash || skb->sw_hash != false) {
 		AIROHA_LOG(AIROHA_DEBUG_LEVEL_INFO, "Incorrect hash:%u\n", hash);
-        return;
-    }
-	
+		return;
+	}
+
 	spin_lock_bh(&ppe_lock);
 	hwe = airoha_ppe_foe_get_entry_locked(ppe, hash);
 	if (!hwe)
@@ -10442,10 +10441,10 @@ int arht_ppe_multicast_handler(struct ai
 {
 	unsigned int foe_index = FOE_ENTRY_NUM(skb);
 
-	if(foe_index >= PPE_SRAM_NUM_ENTRIES || !skb->l4_hash || skb->sw_hash != false)
+	if (foe_index >= airoha_ppe_get_total_sram_num_entries(ppe) || !skb->l4_hash || skb->sw_hash != false)
 		return 0;
-	
-	if(arht_multicast_hwnat_data_handler_hook)
+
+	if (arht_multicast_hwnat_data_handler_hook)
 	{
 		struct airoha_foe_entry *hwe = NULL;
 		spin_lock_bh(&ppe_lock);
@@ -10485,12 +10484,9 @@ int arht_ppe_multicast_set_valid(struct
 {
 	unsigned int foe_index = FOE_ENTRY_NUM(skb);
 
-	if(arht_multicast_hwnat_set_valid_hook)
-	{
-		if(foe_index < PPE_SRAM_NUM_ENTRIES)
-		{
+	if (arht_multicast_hwnat_set_valid_hook) {
+		if (foe_index < airoha_ppe_get_total_sram_num_entries(glb_eth->ppe))
 			arht_multicast_hwnat_set_valid_hook(foe_index);
-		}
 	}
 	return 0;
 }
@@ -11780,12 +11776,9 @@ int arht_ppe_soe_offload_get_valid(struc
 {
 	unsigned int foe_index = FOE_ENTRY_NUM(skb);
 
-	if(arht_soe_offload_get_valid_hook)
-	{
-		if(foe_index < PPE_SRAM_NUM_ENTRIES)
-		{
+	if (arht_soe_offload_get_valid_hook) {
+		if (foe_index < airoha_ppe_get_total_sram_num_entries(glb_eth->ppe))
 			return arht_soe_offload_get_valid_hook(foe_index);
-		}
 	}
 	return 0;
 }
@@ -11802,13 +11795,11 @@ int airoha_ppe_drop_packet_handler(struc
 		return 0;
 	ppe = glb_eth->ppe;
 
-	if(!airoha_is_pon_sfu_mode()){
+	if (!airoha_is_pon_sfu_mode())
 		return 0;
-	}
 
-	if (hash >= PPE_SRAM_NUM_ENTRIES || !skb->l4_hash || skb->sw_hash != false) {
-        return 0;
-    }
+	if (hash >= airoha_ppe_get_total_sram_num_entries(ppe) || !skb->l4_hash || skb->sw_hash != false)
+		return 0;
 
 	spin_lock_bh(&ppe_lock);
 	hwe = airoha_ppe_foe_get_entry_locked(ppe, hash);
@@ -12788,9 +12779,9 @@ int arht_general_offload_bind(struct sk_
 	int ring;
 
 	foe_entry_idx = skb->hash;
-	if (foe_entry_idx >= PPE_SRAM_NUM_ENTRIES || !skb->l4_hash || skb->sw_hash != false)
+	if (foe_entry_idx >= airoha_ppe_get_total_sram_num_entries(glb_eth->ppe) || !skb->l4_hash || skb->sw_hash != false)
 		return ret;
-	
+
 	spin_lock_bh(&ppe_lock);
 	hwe = airoha_ppe_foe_get_entry_locked(glb_eth->ppe, foe_entry_idx);
 	if (!hwe)
--- a/drivers/net/ethernet/airoha/airoha_ppe.c
+++ b/drivers/net/ethernet/airoha/airoha_ppe.c
@@ -66,7 +66,7 @@ static int airoha_ppe_get_total_num_stat
 	return num_stats;
 }
 
-static u32 airoha_ppe_get_total_sram_num_entries(struct airoha_ppe *ppe)
+u32 airoha_ppe_get_total_sram_num_entries(struct airoha_ppe *ppe)
 {
 	struct airoha_eth *eth = ppe->eth;
 
